File: extension/core/src/main/java/io/holunda/camunda/bpm/data/reader/CaseServiceVariableReader.java
package io.holunda.camunda.bpm.data.reader;

import io.holunda.camunda.bpm.data.factory.VariableFactory;
import org.camunda.bpm.engine.CaseService;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;
import java.util.Optional;

/**
 * Allows reading multiple variable values from {@link CaseService#getVariable(String, String)}.
 */
public class CaseServiceVariableReader implements VariableReader {

  private final CaseService caseService;
  private final String caseExecutionId;

  /**
   * Constructs a reader operating on execution.
   *
   * @param caseService     runtime service to use.
   * @param caseExecutionId execution id.
   */
  public CaseServiceVariableReader(CaseService caseService, String caseExecutionId) {
    this.caseService = caseService;
    this.caseExecutionId = caseExecutionId;
  }

  @NotNull
  @Override
  public <T> Optional<T> getOptional(VariableFactory<T> variableFactory) {
    return variableFactory.from(caseService, caseExecutionId).getOptional();
  }

  @NotNull
  @Override
  public <T> T get(VariableFactory<T> variableFactory) {
    return variableFactory.from(caseService, caseExecutionId).get();
  }

  @NotNull
  @Override
  public <T> T getLocal(VariableFactory<T> variableFactory) {
    return variableFactory.from(caseService, caseExecutionId).getLocal();
  }

  @NotNull
  @Override
  public <T> Optional<T> getLocalOptional(VariableFactory<T> variableFactory) {
    return variableFactory.from(caseService, caseExecutionId).getLocalOptional();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    CaseServiceVariableReader that = (CaseServiceVariableReader) o;

    if (!Objects.equals(caseService, that.caseService)) return false;
    return Objects.equals(caseExecutionId, that.caseExecutionId);
  }

  @Override
  public int hashCode() {
    int result = caseService != null ? caseService.hashCode() : 0;
    result = 31 * result + (caseExecutionId != null ? caseExecutionId.hashCode() : 0);
    return result;
  }
}

File: extension/core/src/main/kotlin/io/holunda/camunda/bpm/data/reader/CaseServiceVariableReader.kt
package io.holunda.camunda.bpm.data.reader

import io.holunda.camunda.bpm.data.factory.VariableFactory
import org.camunda.bpm.engine.CaseService
import java.util.*

/**
 * Allows reading multiple variable values from [CaseService.getVariable].
 * @param caseService     runtime service to use.
 * @param caseExecutionId execution id.
 *
 */
class CaseServiceVariableReader(
  private val caseService: CaseService,
  private val caseExecutionId: String
) : VariableReader {

    override fun <T> getOptional(variableFactory: VariableFactory<T>): Optional<T> {
        return variableFactory.from(caseService, caseExecutionId).getOptional()
    }

    override fun <T> get(variableFactory: VariableFactory<T>): T {
        return variableFactory.from(caseService, caseExecutionId).get()
    }

    override fun <T> getLocal(variableFactory: VariableFactory<T>): T {
        return variableFactory.from(caseService, caseExecutionId).getLocal()
    }

    override fun <T> getLocalOptional(variableFactory: VariableFactory<T>): Optional<T> {
        return variableFactory.from(caseService, caseExecutionId).getLocalOptional()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as CaseServiceVariableReader
        return if (caseService != that.caseService) false else caseExecutionId == that.caseExecutionId
    }

    override fun hashCode(): Int {
      return Objects.hash(caseService, caseExecutionId)
    }
}
